### PR TITLE
Add doc comments

### PR DESCRIPTION
**Pull request by jwood803**
Fix for #1265

Currently marked as WIP since I'm sure the doc comment summaries may need to be updated to provide more information. Also, the `netstandard` version of the `CpuMathUtils` will also need to be updated with doc comments.

### CODE REVIEW

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 1]
```diff
@@ -24,6 +24,14 @@ public static partial class CpuMathUtils
         public static int GetVectorAlignment()
             => Avx.IsSupported ? Vector256Alignment : (Sse.IsSupported ? Vector128Alignment : FloatAlignment);
 
+        /// <summary>
+        /// Multiplies a matrix times a source.
+        /// </summary>
+        /// <param name="transpose"></param>
```
Comment (sfilipi) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: If you leave the params here, you'd need to put some explanation text on them.

Comment (jwood803) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: @sfilipi The params are now documented. Though, I'm sure some of them don't offer much value.

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 2]
```diff
@@ -701,6 +853,15 @@ private static unsafe void ZeroMatrixItemsCore(AlignedArray destination, int c,
...
         }
 
+        /// <summary>
+        /// Updates span items with threshold.
+        /// </summary>
+        /// <param name="primalUpdate">The primal update value.</param>
+        /// <param name="count">The count of items.</param>
+        /// <param name="source">The source values.</param>
+        /// <param name="threshold">The threshold value.</param>
+        /// <param name="v">The v span.</param>
```
Comment (jwood803) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Wasn't sure what this was exactly so the param documentation is very simple.

Also, should this be renamed since this file is being updated?

**Comment by shauheen** [comment]
@jwood803 thanks for your work. Are you planning to complete this PR?

**Comment by jwood803** [comment]
I'll work on this more, as well, within the next few days.


________________________________

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 3]
```diff
@@ -24,6 +24,14 @@ internal static partial class CpuMathUtils
         public static int GetVectorAlignment()
             => Avx.IsSupported ? Vector256Alignment : (Sse.IsSupported ? Vector128Alignment : FloatAlignment);
 
+        /// <summary>
+        /// Multiplies a matrix times a source.
+        /// </summary>
+        /// <param name="transpose">Boolean to transpose the matrix or not.</param>
```
Comment (sharwell) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: 💡 I would typically write this using the following language:

```suggestion
        /// <param name="transpose"><see langword="true"/> to transpose the matrix; otherwise, <see langword="false"/>.</param>
```

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 4]
```diff
@@ -200,7 +236,13 @@ public static void Scale(float value, ReadOnlySpan<float> source, Span<float> de
             }
         }
 
-        // destination[i] = scale * (destination[i] + addend)
+        /// <summary>
+        /// Add to the destination by scale with an addend value.
+        /// destination[i] = scale * (destination[i] + addend)
```
Comment (sharwell) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]:
```suggestion
        /// <code>
        /// destination[i] = scale * (destination[i] + addend)
        /// </code>
```

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 5]
```diff
@@ -382,6 +467,11 @@ public static void MulElementWise(ReadOnlySpan<float> left, ReadOnlySpan<float>
             }
         }
 
+        /// <summary>
+        /// Sum the values in the source.
+        /// </summary>
+        /// <param name="source">The source values.</param>
+        /// <returns>The sum of all items in the source.</returns>
```
Comment (sharwell) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: :bulb:

```suggestion
        /// <returns>The sum of all items in <paramref name="source"/>.</returns>
```

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 6]
```diff
@@ -405,6 +495,11 @@ public static float Sum(ReadOnlySpan<float> source)
             }
         }
 
+        /// <summary>
+        /// Sum the squares of each item in the source.
+        /// </summary>
+        /// <param name="source">The source values.</param>
+        /// <returns>The sum of all items squared in the source.</returns>
```
Comment (sharwell) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: 💡 

```suggestion
        /// <returns>The sum of the squares of all items in <paramref name="source"/>.</returns>
```

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 7]
```diff
@@ -428,6 +523,12 @@ public static float SumSq(ReadOnlySpan<float> source)
             }
         }
 
+        /// <summary>
+        /// Sum the square of each item in the source and subtract the mean.
+        /// </summary>
+        /// <param name="mean">The mean value.</param>
+        /// <param name="source">The source values.</param>
+        /// <returns>The sum of all items in the source by the mean.</returns>
```
Comment (sharwell) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: 😕 I would reword this one.

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs (hunk 3 above)
Comment (TomFinley) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: TIL there's another attribute on the `see` element besides `href`...

**Comment by shauheen** [comment]
Hi @jwood803, this PR is getting rather old, would you have time to complete it?

**Comment by jwood803** [comment]
Yes, though I've been having trouble building the solution locally to address the CI errors and haven't been able to get past it yet.


________________________________

**Comment by shauheen** [comment]
Please let us know if you need help with the builds.

**Comment by jwood803** [comment]
I got the build working, but I'm not sure why tests fail on .NET Core 3 in CI. Running locally everything seems to work.

Also, removing the WIP to make this an official PR, but I'll update it if any changes are needed.

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 8]
```diff
@@ -89,7 +97,18 @@ public static void MatrixTimesSource(bool transpose, AlignedArray matrix, Aligne
...
+        /// </summary>
+        /// <param name="matrix">The input matrix.</param>
+        /// <param name="rgposSrc">The source positions.</param>
+        /// <param name="sourceValues">The source values.</param>
+        /// <param name="posMin">The minimum position.</param>
+        /// <param name="iposMin">The minimum position index.</param>
+        /// <param name="iposLimit">The position limit.</param>
+        /// <param name="destination">The destination matrix.</param>
+        /// <param name="stride">The column stride.</param>
+        public static void MatrixTimesSource(AlignedArray matrix, int[] rgposSrc, AlignedArray sourceValues,
```
Comment (Anipik) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: The failure in the 3.0 ci leg is here. you changed the ```ReadOnlySpan<int>``` to ```int[]``` can you please revert it back ?

Comment (Anipik) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: np @jwood803 :)

Comment (danmosemsft) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Thanks @anipik I missed that. Proves the value of this new CI leg.

**Comment by shauheen** [comment]
@sharwell , @Anipik would you be able to take a final look now that this PR is passing all builds. Thanks.

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 9]
```diff
@@ -25,6 +25,14 @@ internal static partial class CpuMathUtils
         public static int GetVectorAlignment()
             => Avx.IsSupported ? Vector256Alignment : (Sse.IsSupported ? Vector128Alignment : FloatAlignment);
 
+        /// <summary>
+        /// Multiplies a matrix times a source.
+        /// </summary>
+        /// <param name="transpose"><see langword="true"/> to transpose the matrix; otherwise <see langword="false"/>.</param>
```
Comment (sharwell) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Overall this is exactly the kind of comment I would like to see for a parameter. There are two changes I would suggest:

1. Using a parameter reference so it's clear which matrix will be transposed (this is my bad for not being clearer earlier)
2. Add a comma to match the form seen throughout docs.microsoft.com

```suggestion
        /// <param name="transpose"><see langword="true"/> to transpose <paramref name="matrix"/>; otherwise, <see langword="false"/>.</param>
```

Comment (danmosemsft) [reply src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: Since this PR is old and a net improvement I suggest merging as is and this can be done subsequently

**Comment by danmosemsft** [comment]
Thanks @jwood803! At present the team is working to add as many missing XML doc tags as possible before release. If you are interested in continuing that would be wonderful.

**Comment by jwood803** [comment]
I would be glad to! Is there an area that needs the most attention at the moment?

Thank you for letting me help out!


________________________________
